037-prototypes/proposal.md

```diff
@@ -599,8 +599,325 @@ prototypes with special pipeline pipeline semantics and step syntax. These
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "files": {
+      "type": "array",
+      "items": {
+        "type": "string",
+        "concourse:input": {
+          "name": "((name))"
+        }
+      }
+    },
+    "other_config": {
+      "type": "string"
+    },
+    "output_as": {
+      "type": "string",
+      "concourse:output": {
+        "name": "((name))"
+      }
+    }
+  },
+  "required": ["context"]
+}
+```
+
+This isn't fully fleshed out, but the key concept is these `concourse:*`
+keywords, which allow you to say: this element in the object is an
+input/output.
+
+**Pros**
+
+* Performance implications from 1a disappear - much easier to cache
+* All inputs/outputs need to be mentioned in the config, so nothing is implicit
+  * Can also be viewed as a Con - more verbose, and is closer to the original
+  * Unlike the `inputs`/`input_mapping`, `outputs`/`output_mapping` approach,
+    however, this approach requires less repetition, as the inputs/outputs only
+    need to be defined in one place
+* Gives a way for Concourse to easily validate input to a prototype, and for
+  IDE's to provide more useful auto-complete suggestions - related to
+  https://github.com/concourse/concourse/issues/481
+  * The IDE aspect may not be practical as getting the JSON schema requires
+    making the `info` request against a Docker image.
+
+**Cons**
+
+* Makes it much more of a burden to write prototypes, as each message *needs* a
+  JSON schema
+* Probably too restrictive - requires Concourse to support specific custom
+  keywords in the JSON schema. If you wanted to define a map of inputs to
+  paths, for instance, we'd need to provide a special keyword for that (as
+  `concourse:input` alone isn't enough)
+
+#### Option 2a - emit outputs in response objects
+
+This option implies two things:
+
+1. A way for pipelines to explicitly specify what inputs should be provided to
+   a prototype (whereas options 1a/b had the prototypes telling Concourse what
+   inputs should be provided)
+2. A way for prototypes to provide output artifacts as part of its message
+   response (whereas option 1a included in the info response, i.e. not at
+   "runtime" w.r.t. running the message)
+
+The response objects could look something like:
+
+```json
+{
+  "object": {
+    "some_data": 123,
+    "some_output": {"artifact": "./path"}
+  }
+}
+```
+
+Concourse will interpret `{"artifact": "./path"}` as an output, where `./path`
+is relative to some path that's mounted by default to all prototypes. *This
+means that multiple outputs may share an output volume, and differ only by the
+path within that volume.*
+
+Since prototypes can emit multiple response objects, this also means you can
+have *streams* of outputs sharing a name that are identified by some other
+field(s). e.g. the above artifact could be identified as
+`some_output{some_data: 123}` (or something along those lines). That gives you
```

Comment:
I can see use cases for emitting streams of outputs (across `MessageResponses`) - for instance, you have a Go prototype that can compile multiple packages simultaneously, and you want each package to have its own artifact - or you have a code scanner prototype that can scan multiple repos simultaneously.

However, in these cases, it's probably not uncommon to want single standalone outputs as well. For the `go build` prototype, for instance, you'd probably want to cache the module cache in the `GOPATH`. Since all packages would be built using the same `GOPATH`, there's really only a single `modcache` output - not one per `MessageResponse`. Forcing the outputs to be defined alongside streams of data make this a bit awkward - should the same artifact be emitted in each of the `MessageResponses`? Just the first?

---

Granted, the use cases I mentioned are mainly around performance (not wanting to spin up many containers, and instead shoehorning some of the `across` step's behaviour into prototypes). tbh don't have any strong use cases in mind for streams of outputs that can't be accomplished by having a stream of `run` steps instead, at the cost of performance.